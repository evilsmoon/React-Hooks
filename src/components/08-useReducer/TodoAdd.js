import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({ handelAddTodo }) => {

    const [{ description }, handelInputChange, reset] = useForm({
        description: ""
    })

    const handelSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }

        const newToDo = {
            id: new Date().getTime(),
            desc: description,
            done: false

        }
        handelAddTodo(newToDo);
        reset();

    }
    return (

        <>
            <h4>add ToDo</h4>
            <hr />
            <form onSubmit={handelSubmit}>
                <input
                    type='text'
                    name='description'
                    placeholder='add..'
                    autoComplete='off'
                    className='form-control'
                    value={description}
                    onChange={handelInputChange}
                />
                <button
                    type='submit'
                    className='btn btn-primary mt-1 btn-block'
                >
                    Add
               </button>
            </form>
        </>
    )
}
