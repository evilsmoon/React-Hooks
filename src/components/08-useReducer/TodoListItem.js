import React from 'react'
import './styles.css'

export const TodoListItem = ({ todo, index, handelToggle ,handelDelete}) => {
    // console.log(todo, i)
    return (
        <li
            key={todo.id}
            className='list-group-item'
        >
            <p
                className={`${todo.done && 'complete'} `}
                onClick={() => {
                    handelToggle(todo.id)

                }}
            >
                {index + 1}. {todo.desc}
            </p>
            <button
                className='btn btn-danger'
                onClick={() => {
                    handelDelete(todo.id)
                }}
            >
                delete
            </button>
        </li>
    )
}
