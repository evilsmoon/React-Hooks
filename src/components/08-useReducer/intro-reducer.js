
const initialState =[{
    id:1,
    todo:'buy weed',
    done:false
}]

const todoReducer = (state = initialState,action)=>{
   
   if(action?.type === 'add'){
       
       return [...state,action.payload];
   }
   
}
let todos = todoReducer();

// todos.push() no usar push en los arreglos por que muta

const newTodo = {
    id:2,
    todo:'smook weed',
    done:false
}

const addTodoAction = {
    type:'add',
    payload: newTodo
}
todos = todoReducer( todos,addTodoAction );
console.log(todos)

