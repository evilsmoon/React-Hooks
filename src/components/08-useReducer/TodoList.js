import React from 'react';
import { TodoListItem } from './TodoListItem';
export const TodoList = ({ todos ,handelDelete,handelToggle}) => {

    console.log(todos)
    return (
        <ul className='list-group list-group-flush'>
            {
                todos.map((todo, i) => (
                    
                    <TodoListItem 
                    key={ todo.id }
                    todo={ todo } 
                    index={ i } 
                    handelToggle={handelToggle} 
                    handelDelete={handelDelete}/>
                    
                ))
            }

        </ul>
    )
}

// TodoList.propTypes = {

// }

