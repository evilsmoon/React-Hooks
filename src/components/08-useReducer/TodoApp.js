import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';
// import { useForm } from '../../hooks/useForm';
import { TodoList } from './TodoList';

import './styles.css'
import { TodoAdd } from './TodoAdd';

const init = ()=> {
    // return [{
    // id: new Date().getTime(),
    // desc: 'aprendiendo React',
    // done : false
    // }];
    return JSON.parse(localStorage.getItem('todos')) || [];
}
export const TodoApp = () => {
    
    
    const [todos,dispatch] = useReducer(todoReducer, [],init);
    
    

    useEffect(() => {
        localStorage.setItem('todos',JSON.stringify(todos));
    }, [todos]);


    const handelDelete = ( todoId ) =>{
        const action = {
            type: 'delete',
            payload:todoId 

            
        }
        dispatch(action);
   
    }

    const handelToggle = ( todoId ) =>{

        dispatch({
            type:'toggle',
            payload:todoId
        })
    }

    const handelAddTodo = ( newToDo) =>{
        dispatch({
            type: 'add',
            payload: newToDo
        } )
    }
    

    return (
        <div>
            <h1>Todo <small>{todos.length}</small></h1>
            <hr/>
        <div className='row'>
            <div className='col-7'>
            <>
            <TodoList 
                    todos={todos} 
                    handelDelete={handelDelete} 
                    handelToggle={handelToggle}/>
            </>

            </div>
            <div className='col-5'>
                <TodoAdd 
                    handelAddTodo={handelAddTodo}/>
            </div>
        </div>
       
        </div>
    )
}
