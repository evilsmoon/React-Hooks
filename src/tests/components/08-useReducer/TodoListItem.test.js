import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import React from 'react';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from './fixtures/demoTodos';


describe('test TodoListItem.js', () => {

    const handelToggle = jest.fn();
    const handelDelete = jest.fn();

    const wrapper = shallow(
        <TodoListItem
            todo={demoTodos[0]}
            index={0}
            handelToggle={handelToggle}
            handelDelete={handelDelete}
        />
    );

    test('debe de mostrar correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('debe de llamar la funcion handelToggle', () => {

        wrapper.find('p').simulate('click')
        expect(handelToggle).toHaveBeenCalledTimes(1)
        expect(handelToggle).toHaveBeenCalledWith(demoTodos[0].id)




    });
    test('debe de llamar la funcion handelDelete', () => {
        wrapper.find('button').simulate('click')
        expect(handelDelete).toHaveBeenCalledTimes(1)
        expect(handelDelete).toHaveBeenCalledWith(demoTodos[0].id)

    });
    test('debe de mostrar el text correctamente', () => {

        const p = wrapper.find('p');
        // console.log(p.text())
        expect(p.text().trim()).toBe(`1. ${demoTodos[0].desc}`)
    });

    test('debe de mostrar la clase del parrafo', () => {


        const todo = demoTodos[0];
        todo.done = true
        const wrapper = shallow(
            <TodoListItem
                todo={todo}
            />
        );
        expect(wrapper.find('p').hasClass('complete')).toBe(true);


    })


})
