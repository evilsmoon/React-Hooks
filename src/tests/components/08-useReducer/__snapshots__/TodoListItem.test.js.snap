// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test TodoListItem.js debe de mostrar correctamente 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false "
    onClick={[Function]}
  >
    1
    . 
    aprender react
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    delete
  </button>
</li>
`;
